core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -165,24 +170,62 @@ private void acquire(SessionProtocol desiredProtocol, PoolKey key,
             return;
         }
 
-        // Try to use the pending connection to avoid creating an extra connection.
-        final Future<Channel> pendingChannel = getPendingConnection(desiredProtocol, key);
-        if (pendingChannel != null) {
-            // Acquire again after the pending connection is completed.
-            pendingChannel.addListener(unused -> acquire(desiredProtocol, key, promise));
+        if (usePendingAcquisition(desiredProtocol, key, promise)) {
             return;
         }
+        setPendingAcquisition(desiredProtocol, key, promise);
 
         // Create a new connection.
         final Future<Channel> f = connect(desiredProtocol, key);
-        setPendingConnection(desiredProtocol, key, f);
         if (f.isDone()) {
             notifyConnect(desiredProtocol, key, f, promise);
         } else {
             f.addListener((Future<Channel> future) -> notifyConnect(desiredProtocol, key, future, promise));
         }
     }
 
+    /**
+     * Tries to use the pending HTTP/2 connection to avoid creating an extra connection.
+     *
+     * @return {@code true} if succeeded to reuse the pending connection.
+     */
+    private boolean usePendingAcquisition(SessionProtocol desiredProtocol, PoolKey key,
+                                          CompletableFuture<PooledChannel> promise) {
+
+        if (desiredProtocol == SessionProtocol.H1 || desiredProtocol == SessionProtocol.H1C) {
+            // Can't use HTTP/1 connections because they will not be available in the pool until
+            // the request is done.
+            return false;
+        }
+
+        final CompletableFuture<PooledChannel> pendingAcquisition =
+                getPendingAcquisition(desiredProtocol, key);
+
+        if (pendingAcquisition == null) {
+            return false;
+        }
+
+        pendingAcquisition.handle((pch, cause) -> {
+            if (cause == null) {
+                if (pch.protocol().isMultiplex()) {
+                    promise.complete(pch);
+                } else {
+                    // Try to acquire again because the connection was not HTTP/2.
+                    // We use the exact protocol (H1 or H1C) instead of 'desiredProtocol' so that
+                    // we do not waste our time looking for pending acquisitions for the host
+                    // that does not support HTTP/2.
+                    acquire(pch.protocol(), key, promise);
+                }
+            } else {
+                // Try to acquire again because the pending connection attempt has failed.
+                acquire(desiredProtocol, key, promise);
```

Comment:
Might be silly question, but just wondering if there's any chance that an event loop tries this `acquire` forever. :-)

Reply:
I guess it is possible. Pushed a fix.